File: 2026-04/src/day_14/boj_1201/Main.java
package day_14.boj_1201;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        if ((n + k - 1) / k > m || n - k + 1 < m) {
            System.out.println(-1);
            return;
        }

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = i + 1;
        }

        for (int i = 0; i < n; i += k) {
            reverse(arr, i, Math.min(i + k - 1, n - 1));
        }

        int cnt = m - (n + k - 1) / k;
        int right = n - 1;
        while (cnt > 0) {
            int max = 0;
            int idx = 0;
            for (int i = 0; i <= right; i++) {
                if (arr[i] > max) {
                    max = arr[i];
                    idx = i;
                }
            }

            for (int i = idx; i < right; i++) {
                arr[i] = arr[i + 1];
            }
            arr[right] = max;

            if (idx != right) cnt--;
            right--;
        }

        for (int x : arr) {
            sb.append(x).append(" ");
        }

        System.out.println(sb);
    }

    static void reverse(int[] arr, int start, int end) {
        while (start < end) {
            int tmp = arr[start];
            arr[start] = arr[end];
            arr[end] = tmp;
            start++;
            end--;
        }
    }
}

File: 2026-04/src/day_14/boj_1201/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m, k;
    cin >> n >> m >> k;

    if ((n + k - 1) / k > m || n - k + 1 < m) {
        cout << -1;
        return 0;
    }

    vector<int> v(n);
    iota(v.begin(), v.end(), 1);

    for (int i = 0; i < n; i += k) {
        reverse(v.begin() + i, v.begin() + min(i + k, n));
    }

    int cnt = m - (n + k - 1) / k;
    int right = n - 1;
    while (cnt > 0) {
        int max = 0;
        int idx = 0;
        for (int i = 0; i <= right; i++) {
            if (v[i] > max) {
                max = v[i];
                idx = i;
            }
        }

        for (int i = idx; i < right; i++) {
            v[i] = v[i + 1];
        }
        v[right] = max;

        if (idx != right) cnt--;
        right--;
    }

    for (int x : v) {
        cout << x << ' ';
    }
}
